cpp/proto/serializer_v2.cc

```diff
@@ -0,0 +1,108 @@
+/* -*- indent-tabs-mode: nil -*- */
+#include "proto/serializer_v2.h"
+
+#include <glog/logging.h>
+
+#include "util/status.h"
+#include "util/openssl_util.h"
+
+namespace rfc6962_bis {
+using util::StatusOr;
+using util::Status;
+
+namespace {
+const char kDerOIDTag = 0x06;
+}
+
+OID::OID() : oid_(nullptr) {
+}
+
+OID::OID(ASN1_OBJECT* oid) : oid_(oid) {
+  CHECK_NOTNULL(oid_);
+}
+
+OID::OID(const OID& other) : oid_(nullptr) {
+  if (other.oid_ != nullptr) {
+    oid_ = OBJ_dup(other.oid_);
+  }
+}
+
+OID::~OID() {
+  if (oid_ != nullptr) {
+    ASN1_OBJECT_free(oid_);
+  }
+}
+
+std::string OID::ToTagMissingDER() const {
+  if(oid_ == nullptr) {
+    // Uninitialized.
+    return "";
+  }
+  int encoded_length = i2d_ASN1_OBJECT(oid_, NULL);
+  if (encoded_length <= 0) {
+    LOG(WARNING) << "Cannot encode OID, OpenSSL error: "
+                 << util::DumpOpenSSLErrorStack();
+    return "";
+  }
+
+  unsigned char* encoded_oid =
+      static_cast<unsigned char*>(OPENSSL_malloc(encoded_length));
+  // i2d_ASN1_OBJECT will change the pointer, so have to use a temporary.
+  unsigned char* tmp_ptr = encoded_oid;
+  encoded_length = i2d_ASN1_OBJECT(oid_, &tmp_ptr);
+  if (encoded_length <= 0) {
+    LOG(WARNING) << "Cannot encode OID, OpenSSL error: "
+                 << util::DumpOpenSSLErrorStack();
+    return "";
+  }
+
+  // If the first byte, the tag, does not indicate an OID, then OpenSSL was not
+  // used correctly.
+  CHECK_EQ(encoded_oid[0], kDerOIDTag);
+  // Skip the tag byte.
+  std::string out(reinterpret_cast<char*>(encoded_oid + 1), encoded_length - 1);
+  OPENSSL_free(encoded_oid);
+  return std::move(out);
+}
+
+std::string OID::ToString() const {
+  if (oid_ == nullptr) {
+    return "";
+  }
+
+  const size_t kOutputSize = 100;
```

Comment:
OpenSSL recommends 80 - so I've changed and added a reference to the manpage recommending the size (https://www.openssl.org/docs/manmaster/crypto/OBJ_obj2txt.html). Also moved to the top level.